Make operators with optional Tensor? arguments c10-full

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42043 Remove argument unpacking from variable kernels
* #42025 IValue stores undefined tensor as None
* #41947 Change C++ frontend to take optional<Tensor> arguments
* **#41610 Make operators with optional Tensor? arguments c10-full**

Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

smessmer committed:
Make operators with optional Tensor? arguments c10-full

Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take case of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take case of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take case of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -360,6 +360,10 @@ Therefore, for the moment, this is all copy pasted in from VariableTypeEverythin
   m.impl(REGISTER_NAME, \
     &WrapFunction<CastPolicy::POLICY, SIGNATURE, SIGNATURE, &FUNC>::type::call);
 
+#define KERNEL_HACKY_WRAPPER_FOR_LEGACY_SIGNATURES(FUNC, REGISTER_NAME, SIGNATURE, LEGACY_SIGNATURE, POLICY) \
```
Can I pitch `KERNEL_LEGACY_SIGNATURE`? I think that conveys the appropriate level of contingency, without being quite so much of an eyesore (intended, I know! :). It'd be one thing if these were all go away in a week, but it's not quite that ephemeral.

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -360,6 +360,10 @@ Therefore, for the moment, this is all copy pasted in from VariableTypeEverythin
 
+#define KERNEL_LEGACY_SIGNATURE(FUNC, REGISTER_NAME, SIGNATURE, LEGACY_SIGNATURE, POLICY) \
+  m.impl(REGISTER_NAME, \
+    c10::impl::hacky_wrapper_for_legacy_signatures<SIGNATURE>(TORCH_FN((&WrapFunction<CastPolicy::POLICY, LEGACY_SIGNATURE, LEGACY_SIGNATURE, &FUNC>::type::call))));
```
I think this can be directly be eliminated. There aren't N functions you have to port: all you have to do is modify `cached_cast` to handle `optional<Tensor>`. Actually, I'm not sure if AMP properly supported undefined tensors previously anyway (cc @mcarilli )

smessmer replied:
this is eliminated in the next PR stacked on top.

ezyang commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -230,6 +230,22 @@ using supported_primitive_arg_types = guts::typelist::typelist<
+  template<bool AllowDeprecatedTypes>
+  struct ivalue_to_arg<optional<ArrayRef<int64_t>>, AllowDeprecatedTypes> final {
+    // If an argument is optional<ArrayRef<int64_t>>, convert the IValue to a optional<std::vector<int64_t>> and pass that
+    // to the operator.
```
Uhh, so I guess you're just following the same pattern as the code above (which was already doing this), but why do we have to convert the ivalue into a `std::vector<int64_t>`. Isn't this some totally unnecessary allocation?

smessmer replied:
This unboxes an `IValue` (which is a list, so it holds a `std::vector<IValue>`) to an `ArrayRef<int64_t>`. The only way of doing that is to copy the full list. I think I mentioned this a few times when I argued for not allowing `ArrayRef` in kernel signatures, but the resolution was that we said perf doesn't matter here.

ezyang commented on aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:
```diff
@@ -83,13 +84,13 @@ struct with_scattered_tensor_options<UnderlyingFuncPtr, std::enable_if_t<has_ten
         guts::typelist::drop_t<gathered_parameter_types, tensoroptions_arg_index + 1>;
 
-    using wrapper = with_scattered_tensor_options_<UnderlyingFuncPtr, parameters_before_tensoroptions, parameters_after_tensoroptions>;
+    using wrapper = with_scattered_tensor_options__<UnderlyingFuncPtr, parameters_before_tensoroptions, parameters_after_tensoroptions>;
```
double underscore?!

ezyang replied:
(At this point it's probably worth saying why there's two levels of indirection here.)

bhosmer replied:
+1

smessmer replied:
I renamed it to `impl_`

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -80,10 +80,11 @@ namespace {
     bool create_graph) {
-  torch::autograd::backward({self}, {gradient}, std::move(keep_graph), create_graph);
+  Tensor _gradient = gradient.has_value() ? *gradient : Tensor();
+  torch::autograd::backward({self}, {_gradient}, std::move(keep_graph), create_graph);
```
Add a TODO to make `torch::autograd::backward` directly take `c10::optional` here?

ezyang reviewed (approved):
Approving to unblock.

bhosmer commented on aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:
```diff
@@ -105,11 +106,91 @@ struct with_scattered_tensor_options_<FuncPtr, guts::typelist::typelist<Paramete
+ * in those locations, unwraps them to `Tensor` (potentially undefined tensor)
+ * and calls the original kernel function.
+ */
+template<class TargetSignature, class KernelFunc, std::enable_if_t<guts::typelist::true_for_any_type<is_optional_tensor_arg, typename guts::infer_function_traits_t<TargetSignature>::parameter_types>::value, int> = 0>
```
how do you feel about line wrapping above like 100 or 120?

smessmer replied:
generally in favor of clang-formatting everything but we haven't set that up yet

bhosmer commented on aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:
```diff
@@ -105,11 +106,91 @@ struct with_scattered_tensor_options_<FuncPtr, guts::typelist::typelist<Paramete
+ * Take a kernel function that has a number of `Tensor` arguments
+ * and take in a `TargetSignature` that must match, but is allowed
+ * to take `optional<Tensor>` in place of some or all of the `Tensor`
+ * arguments. Returns a new kernel function that has `optional<Tensor>`
```
Personally I'd really benefit from comments on each template definition describing what it does, in addition to a summary comment like this on the public endpoint. E.g. here, everything that's being described (other than the optional Tensor detection) is happening somewhere else, in uncommented templates. Doesn't have to be in depth, just something to label the pieces basically.

smessmer replied:
I added some more comments

bhosmer reviewed (approved):
I'm finding the metaprogramming kind of hard to follow... some comment begging inline :)

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]

smessmer committed:
Update on "Make operators with optional Tensor? arguments c10-full"


Previously, operators that have a `Tensor?` (i.e. optional tensor) in their schema implemented it using `Tensor` in C++ and filled in an undefined tensor for the None case.
The c10 operator library, however, expects `Tensor?` to be represented as `optional<Tensor>`, so those operators couldn't be c10-full yet and still had to use codegenerated unboxing instead of templated unboxing.

This PR changes that. It extends the `hacky_wrapper_for_legacy_signatures` to not only take care of TensorOptions, but now also map between signatures taking `Tensor` and `optional<Tensor>`.
For this, it requires an additional template parameter, the expected signature, and it uses that to go argument-by-argument and unwrap any optionals it finds.

Differential Revision: [D22607879](https://our.internmc.facebook.com/intern/diff/D22607879/)

[ghstack-poisoned]
